dh/sidh/internal/templates/curve_test.gotemp

```diff
@@ -98,3 +102,314 @@ func BenchmarkThreePointLadder(b *testing.B) {
+// We only construct point P and Q because in the attacks the third point is P-Q by construction 
+// and the countermeasure does not test it
+// Without loss of generality, we assume the curve is the starting curve 
+func testInvalidPKFullOrder(t *testing.T) {
+	
+	var P, Q ProjectivePoint
+	
+	P = ProjectivePoint{X: params.B.AffineP, Z: params.OneFp2}
+	Q = ProjectivePoint{X: params.B.AffineQ, Z: params.OneFp2}
+
+	var e3 uint32
+	e3_float := float64(int(params.B.SecretBitLen)+1)/math.Log2(3)
+	e3 = uint32(e3_float)
+
+	rand.Seed(time.Now().UnixNano())
+    random_index_p := rand.Intn(int(e3))
+	random_index_q := rand.Intn(int(e3))
+
+	cparam_q := CalcCurveParamsEquiv3(&params.InitCurve)
+	Pow3k(&P, &cparam_q, uint32(random_index_p))
+	Pow3k(&Q, &cparam_q, uint32(random_index_q))
+
+
+	var invQz, invPz Fp2
+	invQz = Q.Z
+	invPz = P.Z
+	inv(&invQz, &invQz)
+	inv(&invPz, &invPz)
+
+	mul(&P.X, &P.X, &invPz)
+	mul(&Q.X, &Q.X, &invQz)
+	
+	var xP, xQ, xQmP ProjectivePoint
+	xP = ProjectivePoint{X: P.X, Z: params.OneFp2}
+	xQ = ProjectivePoint{X: Q.X, Z: params.OneFp2}
+	xQmP = ProjectivePoint{X: params.OneFp2, Z: params.OneFp2}
+
+	error_verify := PublicKeyValidation(&params.InitCurve, &xP, &xQ, &xQmP, params.B.SecretBitLen) 
+	if error_verify==nil{
+		t.Errorf("\nExpect ciphertext not of full order to fail, index_p: %v  index_q: %v  ", random_index_p, random_index_q)
+	}
+	
+}
+
+// A trivial test case not covered by paper https://eprint.iacr.org/2022/054.pdf and HB
+// Countermeasure in https://eprint.iacr.org/2022/054.pdf only cares about P and Q
+// But if PmQ is point T or O, that can also lead to recovery of the first bit
+func testInvalidPmQ(t *testing.T) {
+
+	var zero Fp2
+	var xP, xQ, xQmP ProjectivePoint
+	xP = ProjectivePoint{X: params.A.AffineP, Z: params.OneFp2}
+	xQ = ProjectivePoint{X: params.A.AffineQ, Z: params.OneFp2}
+	xQmP = ProjectivePoint{X: zero, Z: params.OneFp2}
+
+	error_verify := PublicKeyValidation(&params.InitCurve, &xP, &xQ, &xQmP, params.B.SecretBitLen) 
+	if error_verify==nil{
+		t.Errorf("\nExpect PmQ as T to fail\n")
+	}
+
+}
+
+// Test valid ciphertext
+// Where P, Q are linearly independent points of correct order 3^e3 in E[3^e3]
+func testValidPQ(t *testing.T) {
+
+	var xP, xQ, xQmP ProjectivePoint
+	xP = ProjectivePoint{X: params.B.AffineP, Z: params.OneFp2}
+	xQ = ProjectivePoint{X: params.B.AffineQ, Z: params.OneFp2}
+	xQmP = ProjectivePoint{X: params.OneFp2, Z: params.OneFp2}
+
+	error_verify := PublicKeyValidation(&params.InitCurve, &xP, &xQ, &xQmP, params.B.SecretBitLen) 
+	if error_verify!=nil{
+		t.Errorf("\nExpect correct ciphertext to not fail\n")
+	}
+
+}
+
+
+/* -------------------------------------------------------------------------
+   Public key / Ciphertext validation against attacks proposed in paper https://eprint.iacr.org/2022/054.pdf and HB
+   -------------------------------------------------------------------------*/
+
+func TestInvalidPK(t *testing.T) {
```

Comment:
We can have a separate top-level test for each. Then it's easier to see which one failed, if they fail.

Reply:
sure, sounds good